### PR TITLE
OAK-9120: integrate spellcheck with async iterator

### PR DESCRIPTION
**Pull request by fabriziofortino**
Summary:

* SpellCheck is now completely separated from standard iterator
* Fix index mapping creation: all text fields were made spellcheck-enabled (with .trigram) when at least one spellcheck field was configured
* Async iterator is now the default one
* Removed old sync iterator code

### CODE REVIEW

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticIndexHelper.java [hunk 1]
```diff
@@ -132,25 +143,22 @@ private static void mapInternalProperties(XContentBuilder mappingBuilder) throws
...
             final String name = entry.getKey();
             final List<PropertyDefinition> propertyDefinitions = entry.getValue();
 
-            Type<?> type = Type.fromTag(propertyDefinitions.get(0).getType(), false);
+            Type<?> type = null;
+            boolean useInSpellCheck = false;
+            for (PropertyDefinition pd: propertyDefinitions) {
+                type = Type.fromTag(pd.getType(), false);
+                if (pd.useInSpellcheck) {
+                    useInSpellCheck = true;
```
Comment (thomasmueller) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticIndexHelper.java:157]: + break ?

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticIndexHelper.java [hunk 2]
```diff
@@ -132,25 +143,22 @@ private static void mapInternalProperties(XContentBuilder mappingBuilder) throws
...
-                    "different types " + fields);
-        }
+        checkIndexRules(indexDefinition);
 
         for (Map.Entry<String, List<PropertyDefinition>> entry : indexDefinition.getPropertiesByName().entrySet()) {
             final String name = entry.getKey();
             final List<PropertyDefinition> propertyDefinitions = entry.getValue();
 
-            Type<?> type = Type.fromTag(propertyDefinitions.get(0).getType(), false);
+            Type<?> type = null;
```
Comment (thomasmueller) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticIndexHelper.java:152]: is "type" used later? if not, I would move it into the loop

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/ElasticIndex.java [hunk 3]
```diff
@@ -92,44 +95,38 @@ protected IndexNode acquireIndexNode(String indexPath) {
...
+        final FulltextIndexPlanner.PlanResult planResult = getPlanResult(plan);
+
+        final ElasticRequestHandler requestHandler = new ElasticRequestHandler(plan, planResult);
+        final ElasticResponseHandler responseHandler = new ElasticResponseHandler(planResult, filter);
+
+        final Iterator<FulltextResultRow> itr;
+        if (requestHandler.requiresSpellCheck()) {
+            itr = new ElasticSpellcheckIterator(acquireIndexNode(plan), requestHandler, responseHandler);
+        } else {
+            // this function is called for each extracted row. Passing FulltextIndex::shouldInclude means that for each
```
Comment (thomasmueller) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/ElasticIndex.java:113]: Having such comments is great!

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticIndexHelper.java (hunk 1 above)
Comment (fabriziofortino) [reply oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticIndexHelper.java:157]: good catch!

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticIndexHelper.java (hunk 2 above)
Comment (fabriziofortino) [reply oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticIndexHelper.java:152]: yes, `type` is used by the code after the loop

**Review by averma21 — APPROVED** [review]
LGTM.
Just one question -
> all text fields were made spellcheck-enabled (with .trigram) when at least one spellcheck field was configured

What is the benefit of this change? Not having to reindex if spellcheck is added on a different field later?

**Comment by fabriziofortino** [comment]
@averma21 usually an index definition has just a few spellcheck enabled fields. Spellcheck requires the creation of trigrams that are quite expensive. Creating trigrams for all string fields would result in indexes with an increased size (slow performance both on index and query).
Right now any change in the definition requires a full re-index. We can improve this in the future figuring out what kind od change has been made. We store the `source` in our indexes. In this way we could do a re-index (https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-reindex.html) with the new mapping. This should be quite fast because everything happens in ES (no JCR calls).

**Comment by averma21** [comment]
@fabriziofortino Now i get it. Earlier I thought that you introduced spellchecking for all fields but you have actually removed it. +1 for this.

**Pull request closed without merging**